// scratch-gui/src/components/menu-bar/ModeToggle.jsx
import React, { useState } from 'react';

const ModeToggle = ({ value, onChange }) => {
    const isInteractive = (value === 'interactive');

    const handleToggle = () => {
        const newMode = isInteractive ? 'upload' : 'interactive';
        onChange(newMode); // 通知父组件请求切换
    };

    return (
        <div style={{ display: 'flex', alignItems: 'center' }}>
            <span style={{ marginRight: '10px', color: 'white', fontWeight: 'bold' }}>模式</span>

            <div
                style={{
                    display: 'flex',
                    alignItems: 'center',
                    width: '120px',
                    height: '30px',
                    borderRadius: '999px',
                    backgroundColor: '#00BFFF',
                    // padding: '3px',
                    cursor: 'pointer',
                    position: 'relative',
                }}
                onClick={handleToggle}
            >
                <div
                    style={{
                        position: 'absolute',
                        top: '3px',
                        left: isInteractive ? '3px' : 'calc(50% + 1px)',
                        width: 'calc(50% - 6px)',
                        height: '26px',
                        backgroundColor: 'white',
                        borderRadius: '999px',
                        transition: 'all 0.3s ease',
                        zIndex: 1,
                    }}
                />
                <div
                    style={{
                        flex: 1,
                        textAlign: 'center',
                        color: isInteractive ? '#000' : '#eee',
                        zIndex: 2,
                        fontWeight: 'bold',
                        fontSize: '14px',
                    }}
                >
                    互动
                </div>
                <div
                    style={{
                        flex: 1,
                        textAlign: 'center',
                        color: isInteractive ? '#eee' : '#000',
                        zIndex: 2,
                        fontWeight: 'bold',
                        fontSize: '14px',
                    }}
                >
                    上传
                </div>
            </div>
        </div>
    );
};

export default ModeToggle;
