// scratch-vm/src/util/aiInfo.js
let qrInfo=null

let qrLocation=null

let faceNum=null

let faceLocation=null

let object=null

let objectLocation=null

let trackWhatColor = 'blue'

let isHaveColor=null

let colorLocation=null

let faceName=null

let isSymFace=false

let resultFace=null

let qrWH=[]

let colorWH=[]

let faceWH=[]

let objectWH=[]

let aprilInfo=null

let aprilLocation=null

let aprilWh=null

let region=null

let colorRgb=null

function getQr(){
    return qrInfo
}

function setQr(a){
    qrInfo=a
}

function getQrLocation(){
    return qrLocation
}

function setQrLocation(a){
    qrLocation=a
}

function getFaceNum(){
    return faceNum
}

function setFaceNum(a){
    faceNum=a
}

function getFaceLocation(){
    return faceLocation
}

function setFaceLocation(a){
    faceLocation=a
}

function getObject(){
    return object
}

function setObject(a){
    object=a
}

function getObjectLocation(){
    return objectLocation
}

function setObjectLocation(a){
    objectLocation=a
}

function setWhatColor(a){
    trackWhatColor=a
}

function getWhatColor(){
    return trackWhatColor
}

function setHaveColor(a){
    isHaveColor=a
}

function getHaveColor(){
    return isHaveColor
}

function setColorLocation(a){
    colorLocation=a
}

function getColorLocation(){
    return colorLocation
}

function setFaceName(a){
    faceName=a
}

function getFaceName(){
    return faceName
}

function getIsSym(){
    return isSymFace
}

function setIsSym(a){
    isSymFace=a
}

function setResultFace(a){
    resultFace=a
}

function getResultFace(){
    return resultFace
}

function setQrWh(a){
    qrWH=a
}

function getQrWh(){
    return qrWH
}

function setColorWh(a){
    colorWH=a
}

function getColorWh(){
    return colorWH
}

function setFaceWh(a){
    faceWH=a
}

function getFaceWh(){
    return faceWH
}

function setObjectWh(a){
    objectWH=a
}

function getObjectWh(){
    return objectWH
}

function setAprilInfo(a){
    aprilInfo=a
}

function getAprilInfo(){
    return aprilInfo
}

function setAprilLocation(a){
    aprilLocation=a
}

function getAprilLocation(){
    return aprilLocation
}

function setAprilWh(a){
    aprilWh=a
}

function getAprilWh(){
    return aprilWh
}

function setRegion(a){
    region=a
}

function getRegion(){
    return region
}

function setColorRgb(a){
    colorRgb=a
}

function getColorRgb(){
    return colorRgb
}

module.exports={
    getQr,
    setQr,
    getQrLocation,
    setQrLocation,
    setFaceNum,
    getFaceNum,
    setFaceLocation,
    getFaceLocation,
    setObject,
    getObject,
    setObjectLocation,
    getObjectLocation,
    setWhatColor,
    getWhatColor,
    setHaveColor,
    getHaveColor,
    setColorLocation,
    getColorLocation,
    setFaceName,
    getFaceName,
    getIsSym,
    setIsSym,
    setResultFace,
    getResultFace,
    setQrWh,
    getQrWh,
    setColorWh,
    getColorWh,
    setFaceWh,
    getFaceWh,
    setObjectWh,
    getObjectWh,
    setAprilInfo,
    getAprilInfo,
    setAprilLocation,
    getAprilLocation,
    setAprilWh,
    getAprilWh,
    setRegion,
    getRegion,
    setColorRgb,
    getColorRgb
}
