// scratch-vm/src/util/imageLoad.js
let imageIsLoad=false

let qrIsLoad=false

let faceDetIsLoad=false

let itemIsLoad=false

let colorDete=false

let colorReco=false

let faceReco=false

let isMicphLoad=false

let isAprilTag=false

let isColorPlace = false

function setIsImage(a){
    imageIsLoad=a
}

function getIsImage(){
    return imageIsLoad
}

function setQr(a){
    qrIsLoad=a
}

function getQr(){
    return qrIsLoad
}

function setFaceDet(a){
    faceDetIsLoad=a
}

function getFaceDet(){
    return faceDetIsLoad
}

function setItem(a){
    itemIsLoad=a
}

function getItem(){
    return itemIsLoad
}

function setColorDete(a){
    colorDete=a
}

function getColorDete(){
    return colorDete
}

function setColorReco(a){
    colorReco=a
}

function getColorReco(){
    return colorReco
}

function setFaceReco(a){
    faceReco=a
}

function getFaceReco(){
    return faceReco
}

function getMicph(){
    return isMicphLoad
}

function setMicph(a){
    isMicphLoad=a
}

function getAprilTag(){
    return isAprilTag
}

function setAprilTag(a){
    isAprilTag=a
}

function getColorPlace(){
    return isColorPlace
}

function setColorPlace(a){
    isColorPlace=a
}
module.exports={
    setIsImage,
    getIsImage,
    setQr,
    getQr,
    setFaceDet,
    getFaceDet,
    setItem,
    getItem,
    setColorDete,
    getColorDete,
    setColorReco,
    getColorReco,
    setFaceReco,
    getFaceReco,
    setMicph,
    getMicph,
    getAprilTag,
    setAprilTag,
    getColorPlace,
    setColorPlace
}
