// scratch-vm/src/extension-support/argument-type.js
/**
 * Block argument types
 * @enum {string}
 */
const ArgumentType = {
    /**
     * Numeric value with angle picker
     */
    ANGLE: 'angle',

    /**
     * Boolean value with hexagonal placeholder
     */
    BOOLEAN: 'Boolean',

    /**
     * Numeric value with color picker
     */
    COLOR: 'color',

    /**
     * Numeric value with text field
     */
    NUMBER: 'number',

    /**
     * String value with text field
     */
    STRING: 'string',

    /**
     * String value with matrix field
     */
    MATRIX: 'matrix',

    MATRIXCUSTOM: 'matrix_custom',

    /**
     * MIDI note number with note picker (piano) field
     */
    NOTE: 'note',
  
    /**
     * Inline image on block (as part of the label)
     */
    IMAGE: 'image',

    /**
     * Name of costume in the current target
     */
    COSTUME: 'costume',

    /**
     * Name of sound in the current target
     */
    SOUND: 'sound',

    /**
     * 滑块,限制最大最小输入
     */
    SLIDER: 'slider',
       /**
     *限制输入大小0-4
    */
     NUMRES0_4: 'numres0_4',

     /**
      *限制输入大小0-9
     */
     NUMRES0_9: 'numres0_9',
 
     /**
      *限制输入大小40-500
     */
     NUMRES40_500: 'numres40_500',
 
     /**
      *限制输入大小20-10000
     */
     NUMRES20_10000: 'numres20_10000',
 
     /**
      *限制输入大小0-255
     */
     NUMRES0_255: 'numres0_255',
 
     /**
      *限制输入大小-255-255
     */
     NUMRES_255_255: 'numres_255_255',
 
     /**
      *限制输入大小0-300
     */
     NUMRES0_300: 'numres0_300',
 
     /**
      *限制输入大小-100-100
     */
     NUMRES_100_100: 'numres_100_100',
 
     /**
      *限制输入大小0-
     */
     NUMRES0: 'numres0',
 
     /**
      *限制输入大小0-100
     */
     NUMRES0_100: 'numres0_100',
 
     /**
      *限制输入大小-360-360
     */
     NUMRES_360_360: 'numres_360_360',
 
     /**
      *限制输入大小-32400-32400
     */
     NUMRES_32400_32400: 'numres_32400_32400',
 
     /**
      *限制输入大小0-65535
     */
     NUMRES0_65535: 'numres0_65535',
};

module.exports = ArgumentType;
